src/net/server/dgram.rs

```diff
@@ -0,0 +1,745 @@
+        self.command_tx
+            .lock()
+            .unwrap()
+            .send(ServerCommand::Reconfigure(config))
+            .map_err(|_| Error::CommandCouldNotBeSent)
+    }
+
+    /// Stop the server.
+    ///
+    /// In-flight requests will continue being processed but no new messages
+    /// will be accepted. Pending responses will be written as long as the
+    /// socket that was given to the server when it was created remains
+    /// operational.
+    ///
+    /// [`Self::is_shutdown()`] can be used to dertermine if shutdown is
+    /// complete.
+    ///
+    /// [`Self::await_shutdown()`] can be used to wait for shutdown to
+    /// complete.
+    pub fn shutdown(&self) -> Result<(), Error> {
+        self.command_tx
+            .lock()
+            .unwrap()
+            .send(ServerCommand::Shutdown)
+            .map_err(|_| Error::CommandCouldNotBeSent)
+    }
+
+    /// Check if shutdown has completed.
+    ///
+    /// Note that until shutdown is fully complete some Tokio background tasks
+    /// may remain scheduled or active to process in-flight requests.
+    pub fn is_shutdown(&self) -> bool {
+        self.metrics.num_inflight_requests() == 0
+            && self.metrics.num_pending_writes() == 0
+    }
+
+    /// Wait for an in-progress shutdown to complete.
+    ///
+    /// Returns true if the server shutdown in the given time period, false
+    /// otherwise.
+    ///
+    /// To start the shutdown process first call [`Self::shutdown()`] then use
+    /// this method to wait for the shutdown process to complete.
+    pub async fn await_shutdown(&self, duration: Duration) -> bool {
+        timeout(duration, async {
+            let mut interval = interval(Duration::from_millis(100));
+            interval.set_missed_tick_behavior(MissedTickBehavior::Skip);
+            while !self.is_shutdown() {
+                interval.tick().await;
+            }
+        })
+        .await
+        .is_ok()
+    }
+}
+
+//--- Internal details
+
+impl<Sock, Buf, Svc> DgramServer<Sock, Buf, Svc>
+where
+    Sock: AsyncDgramSock + Send + Sync + 'static,
+    Buf: BufSource + Send + Sync + 'static,
+    Buf::Output: Octets + Send + Sync + 'static,
+    Svc: Service<Buf::Output> + Send + Sync + 'static,
+{
+    /// Receive incoming messages until shutdown or fatal error.
+    ///
+    // TODO: Use a strongly typed error, not String.
+    async fn run_until_error(&self) -> Result<(), String>
+    where
+        Svc::Future: Send,
+    {
+        let mut command_rx = self.command_rx.clone();
+
+        loop {
+            tokio::select! {
+                // Poll futures in match arm order, not randomly.
+                biased;
+
+                // First, prefer obeying `ServerCommand`s over everything
+                // else.
+                res = command_rx.changed() => {
+                    self.process_service_command(res, &mut command_rx)?;
+                }
+
+                recv_res = self.recv_from() => {
```

Comment:
There are two ways to implement rate limiting:
1. During initial datagram receipt. This happens outside the server code in the `AsyncDgramSock` impl supplied by the library user. The default impl we supply for `UdpSocket` doesn't implement any such limiting but a user could add such limiting per their needs at that point by providing their own `AsyncDgramSock` impl for use by the server.
2. Aftger datagram receipt during request pre-processing by middleware. No such rate limiting middleware is implemented in this PR but such a middleware could cease processing if over a limit.

Reply:
I don't think we are asking the application to start implementing UDP stuff. Currently no rate limiting is provided. When we choose to provide it we can either implement it in the supplied `UdpSocket` trait impl, so it's supplied with the library but not in the server code itself, implement in the server code, or implement it in middleware, or even in a service helper, whatever seems best. But rate limiting was not, to my knowledge, in scope for this PR.

Reply:
In my opinion, a proper server transport (bullet point 2) includes rate limiting. As far as I can tell in the current setup the server could run out of memory if too many requests arrive before the service implementation has been able to handle them.

Reply:
I would rather not attempt to implement something quickly now, there are multiple different places it could be implemented with different PROs and CONs and there are different rate limiting algorithms to consider, plus I would want to look at what other DNS server implementations do. So instead I have added rate limiting to our shared TODO list.